app/models/user.rb

```diff
@@ -3,4 +3,28 @@ class User < ActiveRecord::Base
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
+
+  attr_accessor :login
+
+  validates :username,
+            presence: true,
+            uniqueness: {
+              :case_sensitive => false
+            }
+
+  def self.find_first_by_auth_conditions(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions).where([
+        "lower(username) = :value OR lower(email) = :value",
+        { :value => login.downcase }
+        ]).first
```

Comment:
Indent the right bracket the same as the start of the line where the left bracket is.